jb/itch5/compute_book.hpp

```diff
@@ -19,91 +19,96 @@
-
-  /// A convenience typedef for clock_type::time_point
-  typedef typename clock_type::time_point time_point;
+  time_point recvts;
 
   /**
-   * A flat struct to represent updates to an order book.
-   *
-   * Updates to an order book come in many forms, but they can all be
-   * represented with a simple structure that shows: what book is
-   * being updated, what side of the book is being updated, what price
-   * level is being updated, and how many shares are being added or
-   * removed from the book.
+   * The security updated by this order.  This is redundant for
+   * order updates and deletes, and ITCH-5.0 omits the field, but
+   * we find it easier
    */
-  struct book_update {
-    /**
-     * When was the message that triggered this update received
-     */
-    time_point recvts;
+  stock_t stock;
 
-    /**
-     * The security updated by this order.  This is redundant for
-     * order updates and deletes, and ITCH-5.0 omits the field, but
-     * we find it easier
-     */
-    stock_t stock;
+  /// What side of the book is being updated.
+  buy_sell_indicator_t buy_sell_indicator;
 
-    /// What side of the book is being updated.
-    buy_sell_indicator_t buy_sell_indicator;
+  /// What price level is being updated.
+  price4_t px;
 
-    /// What price level is being updated.
-    price4_t px;
+  /// How many shares are being added (if positive) or removed (if
+  /// negative) from the book.
+  int qty;
 
-    /// How many shares are being added (if positive) or removed (if
-    /// negative) from the book.
-    int qty;
+  /// If true, this was a cancel replace and and old order was
+  /// modified too...
+  bool cxlreplx;
 
-    /// If true, this was a cancel replace and and old order was
-    /// modified too...
-    bool cxlreplx;
+  /// Old price for the order
+  price4_t oldpx;
 
-    /// Old price for the order
-    price4_t oldpx;
+  /// How many shares were removed in the old order
+  int oldqty;
+};
 
-    /// How many shares were removed in the old order
-    int oldqty;
-  };
+/**
+  * A convenient container for per-order data.
+  *
+  * Most market data feeds resend the security identifier and side
+  * with each order update, but ITCH-5.0 does not.  One needs to
+  * lookup the symbol, side, original price,  information based on the order
+  * id.  This literal type
+  * is used to keep that information around.
+  */
+struct order_data {
+  /// The symbol for this particular order
+  stock_t stock;
+
+  /// Whether the order is a BUY or SELL
+  buy_sell_indicator_t buy_sell_indicator;
+
+  /// The price of the order
+  price4_t px;
+
+  /// The remaining quantity in the order
+  int qty;
+};
 
+/**
+ * Compute the book and call a user-defined callback on each change.
+ *
+ * Keep a collection of all the order books, indexed by symbol, and
+ * forward the updates to them.  Only process the relevant messages
+ * types in ITCH-5.0 that are necessary to keep the book.
+ */
```

Comment:
Please document the book_type template parameter.  When you do so, please document what are the requirements of the type, not what types it can be used with (you can say "A type with the same interface as 'foo'", please do not say "only use with 'foo'").